### DISCUSSION TITLE
Tabulate differences / progress on operators across StableHLO and its competitors

### DISCUSSION THREAD

**Discussion by SamuelMarks** [openxla/community#128] [stablehlo]
Back in the day this was basically my homepage: https://node.green

The specification in ML world isn't as agreed-upon as [ECMAScript](https://en.wikipedia.org/wiki/ECMAScript); with [BLAS](https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms) and Python-world [NumPy](https://en.wikipedia.org/wiki/NumPy) and [Scikit-learn](https://en.wikipedia.org/wiki/Scikit-learn) having implementations everywhere from [CUDA](https://en.wikipedia.org/wiki/CUDA) ([cuBLAS](https://developer.nvidia.com/cublas), [cuML](https://docs.rapids.ai/api/cuml)) to [Keras](https://en.wikipedia.org/wiki/Keras) to [PyTorch](https://en.wikipedia.org/wiki/PyTorch) to [TensorFlow](https://en.wikipedia.org/wiki/TensorFlow)'s [`tf.experimental.numpy`](https://www.tensorflow.org/api_docs/python/tf/experimental/numpy) to [JAX](https://en.wikipedia.org/wiki/JAX_(software)) to [ArrayFire](https://en.wikipedia.org/wiki/ArrayFire) to [Dask](https://en.wikipedia.org/wiki/Dask_(software)) &etc.

What I would like to see—and happy to assist in its original creation and scripts to maintain it—is a table with these columns:

| Operation | [NumPy](https://numpy.org) | [TensorFlow](https://tensorflow.org) | [JAX](https://jax.dev) | ⋱
| ---------- | ------- | ------------ | -------------- | -------------- |
| [dot](https://en.wikipedia.org/wiki/Dot_product) (⋅) | [✔](https://numpy.org/doc/stable/reference/generated/numpy.dot.html) | [✔](https://www.tensorflow.org/api_docs/python/tf/experimental/numpy/dot) | [✔](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.dot.html) | ︙
| ⋱ | … | … | … | ⋰

_maybe with extra detail like what version it was built, if it is partially implemented, if one impl has `axis` whilst the other has `dimen`, &etc._

It would be great if this was widely featured and placed on each framework's website. The advantage would be encouraging:
- compatibility
- interoperability
- portability
- good TODO lists for engineers that work on each project
- benchmarking different implementations
- understanding how to compare different implementations
  - e.g., one might be better on M [ARM-series] Mac's like [MLX](https://github.com/ml-explore/mlx); whilst another may perform better on TPUs; whereas another still might perform better on—high-end or low-end—AMD or NVIDIA GPUs; &etc.
- developing tooling to automatically convert betwixt implementations… without having to rely on custom-written abstractions over abstractions [adapter-pattern](https://en.wikipedia.org/wiki/Adapter_pattern) ([DI inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle)) frameworks like {[Keras](https://keras.io), [ΦML](https://github.com/tum-pbs/PhiML), [einops](https://github.com/arogozhnikov/einops), …}.
  - Instead one should be able to take an arbitrary PyTorch codebase and convert it to a best-practice following JAX codebase without human intervention (or at least get >95% of the way to this)… and vice-vera.
  - The table could provide sufficient information to convert betwixt formats; and identify gaps in APIs so that a conversion isn't attempted but rather [`NotImplementedError`](https://docs.python.org/3/library/exceptions.html#NotImplementedError) is raised.

FWIW: There is other work in this direction like [Python array API standard](https://data-apis.org/array-api), your [spec](https://github.com/openxla/stablehlo/blob/main/docs/spec.md), [ONNX](https://en.wikipedia.org/wiki/Open_Neural_Network_Exchange), and others. But AFAIK no one is maintaining one of these tables. And maintaining one of these tables outside one of the big projects would be a recipe for obscurity and quickly becoming out-of-date.
